src/Orchard.Web/Modules/Orchard.OpenId/Module.cs

```diff
@@ -0,0 +1,90 @@
+﻿using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Logging;
+using Orchard.Environment.Shell;
+using Orchard.FileSystem.AppData;
+using Orchard.OpenId.Indexes;
+using Orchard.OpenId.Models;
+using Orchard.OpenId.Services;
+using Orchard.Security;
+using Orchard.Settings;
+using Orchard.Users.Models;
+using System;
+
+namespace Orchard.OpenId
+{
+    public class Startup : StartupBase
+    {     
+        private readonly string _certificateFullPath;
+        private const string certificateFileName = "Certificate.pfx";
+        private readonly string tenantUrlPrefix;
+        private readonly string tenantName;
+        public Startup(ShellSettings shellSettings, IAppDataFolder appDataFolder, ILoggerFactory loggerFactory)   
+        {
+            tenantName = shellSettings.Name;
+            tenantUrlPrefix = string.IsNullOrWhiteSpace(shellSettings.RequestUrlPrefix) ? "" : "/" + shellSettings.RequestUrlPrefix;
+            _certificateFullPath = appDataFolder.Combine(shellSettings.Name, certificateFileName);
+        }
+
+        public override void Configure(IApplicationBuilder builder, IRouteBuilder routes, IServiceProvider serviceProvider)
+        {
+            var baseUrl = serviceProvider.GetRequiredService<ISiteService>().GetSiteSettingsAsync().Result.BaseUrl;
+            baseUrl = baseUrl.EndsWith("/") ? baseUrl.Remove(baseUrl.Length - 1) : baseUrl;
+            var tenantUrl = baseUrl + tenantUrlPrefix;
+
+            builder.UseOpenIddict();
+            builder.UseJwtBearerAuthentication(new JwtBearerOptions()
+            {
+                AutomaticAuthenticate = true,
+                AutomaticChallenge = true,
+#if DEBUG
+                RequireHttpsMetadata = false,
+#endif
+                Audience = tenantUrl,
+                Authority = tenantUrl
+            });
+        }
+
+        public override void ConfigureServices(IServiceCollection serviceCollection)
```

Comment:
Per tenant, one if the many reasons to use Orchard for SaaS applications.

On Thu, Aug 11, 2016, 7:14 AM Kévin Chalet <notifications@github.com> wrote:

> In src/Orchard.Web/Modules/Orchard.OpenId/Module.cs
> <https://github.com/OrchardCMS/Orchard2/pull/215#discussion_r74429508>:
>
> > +            var tenantUrl = baseUrl + tenantUrlPrefix;
> > +
> > +            builder.UseOpenIddict();
> > +            builder.UseJwtBearerAuthentication(new JwtBearerOptions()
> > +            {
> > +                AutomaticAuthenticate = true,
> > +                AutomaticChallenge = true,
> > +#if DEBUG
> > +                RequireHttpsMetadata = false,
> > +#endif
> > +                Audience = tenantUrl,
> > +                Authority = tenantUrl
> > +            });
> > +        }
> > +
> > +        public override void ConfigureServices(IServiceCollection serviceCollection)
>
> @sebastienros <https://github.com/sebastienros> is this services
> collection tenant-specific or app-global? If it's shared between tenants,
> then it will be extremely hard to isolate the OpenIddict instances, because
> exactly like ASP.NET Core Identity, it relies on DI-registered options.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/OrchardCMS/Orchard2/pull/215/files/6a4ec2a145817f015613d0f105a2d99d1a775f9a#r74429508>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABHJ7drwLqHJqg_K-jTANLmuzB4sYcTEks5qey42gaJpZM4Jf7LG>
> .
>